### DISCUSSION TITLE
Bundle only used JS and CSS during build

### DISCUSSION THREAD

**Discussion by leifmarcus** [withastro/roadmap#606] [Proposal]
# Summary

As mentioned in different issues already  ([#6563](https://github.com/withastro/astro/issues/6563), [#6906](https://github.com/withastro/astro/issues/6906)) when having a page, that conditionally imports components, all imported components are bundled into one file even if a component is not used on a specific rendered page. As far as I can see, this is mainly relevant for CSS that is bundled in a single file.

# Background & Motivation

I'm working on a marketing website for a large automotive company in Germany. We are currently transition the whole site to jamstack and decided to use the Astro Framework. Everything works quite well and we are happy with the framework so far. 

But, there is one thing that we realized is, that when having an astro page that generates multiple pages and import multiple components into the file, that might not be used on all pages, the css for all the components is bundled in one file. As we have many different components that can potentially be used by content editors, we need kind of a functionality, that maps component names to astro components and based on the api output only render certain components on this pages.

When doing so, the whole CSS is loaded into each of these build pages which is not optimal from a performance point of view.

## Example Implementation

The example shows the implementation with three building blocks for this approach. The idea is, to have a generative approach, that makes it possible to render hundreds of pages with the same page and make it possible for content editors to use whatever content modules they want to use to design the pages.

### 1. define Component Mapping
 
```ts
// index.ts for all potential components on the page
import FirstComponent from './ComponentName/FirstComponent.astro';
import SecondComponent from './ComponentName/SecondComponent.astro';

// could be a map or a object with key, value pairs
export const componentEntries = new Map<string, AstroComponentFactory>([
  // the names are in sync with the names of the components the content editors can place on the pages.
  // with this we map the name of the component to the related astro component.
  ['FirstComponent', FirstComponent]
  ['SecondComponent', SecondComponent]
])
```

### 2. Astro Component as definition for the view of a certain content module component

```astro
---
import FirstComponent from '..../FirstComponent.vue';
const { component } = Astro.props
---
<FirstComponent
  client:visible
  {...component.content}
  config={component.config}
  meta={component.meta}
/>
```

### Usage in Astro Page
The components data is received via the `getStaticPaths` and will get the current components of a given route.

```astro
---
import { componentEntries } from '@/components/index.js'

export async function getStaticPaths(): Promise<GetStaticPathsResult> {
  // return the pages in correct format to be consumed by the build process
}

const { components, meta } = Astro.props
---
<DefaultLayout meta={meta}>
  {
    components.map((component) => {
      const OutputComponent = componentEntries.get(component.name);
      if (OutputComponent) {
        return <OutputComponent component={component} />;
      }
      // return some fallback component else
    })
  }
</DefaultLayout>
```

# Goals

- For each page only required assets for the rendered components should be added to the generated page.
- Alternatively, It would be nice to have an approach how to solve generic pages like this for larger applications.

**Comment by delucis** [comment]
I’ve noticed this too. I think it’s technically tricky to get right because the CSS/JS bundling is done via static analysis (following `import` statements), but whether or not a page _uses_ that is a runtime concern — i.e. a component is imported bringing with it some CSS/JS, and then only when the page is rendered do we know if it is rendered, which is too late for bundling.

It’s also a delicate performance trade-off, because generally you don’t want hundreds of tiny CSS/JS files one for each component as this means many requests and lower compression benefits to load them.

Curious if Astro can improve here.

One approach that can help if you’re looking at static builds, is to use separate dynamic routes for different page types. For example:
```
src/
  pages/
    [...doors].astro
    [...engines].astro
    [...wheels].astro
```
While the same problem persists within each route, this may at least help break up the dependencies of a single route. Of course, if your content can genuinely contain any combination of components, this approach won’t help.

**Reply by leifmarcus** [reply]
Thanks @delucis.

At our product we have multiple teams working on different pages, So we kind of have already a split of these pages through the different applications we have.

I see the problem and that is not easy to solve somehow. But, I just wanted to point it out and maybe there is someone who has an Idea on a solution ;)
